service/matching/tasklist/adaptive_scaler.go

```diff
@@ -177,91 +192,149 @@ func (a *adaptiveScalerImpl) getPartitionConfig() *types.TaskListPartitionConfig
-		a.logger.Info("adjust read partitions", tag.NumReadPartitions(numWritePartitions), tag.NumWritePartitions(numWritePartitions))
-		return numWritePartitions
+func (a *adaptiveScalerImpl) adjustWritePartitions(writePartitions map[int]*types.TaskListPartition, targetWritePartitions int) (map[int]*types.TaskListPartition, bool) {
+	if len(writePartitions) == targetWritePartitions {
+		return writePartitions, false
 	}
-	changed := false
-	// check the backlog of the drained partitions
-	for i := numReadPartitions - 1; i >= numWritePartitions; i-- {
-		resp, err := a.matchingClient.DescribeTaskList(a.ctx, &types.MatchingDescribeTaskListRequest{
-			DomainUUID: a.taskListID.GetDomainID(),
-			DescRequest: &types.DescribeTaskListRequest{
-				TaskListType: a.taskListType,
-				TaskList: &types.TaskList{
-					Name: a.taskListID.GetPartition(int(i)),
-					Kind: types.TaskListKindNormal.Ptr(),
-				},
-				IncludeTaskListStatus: true,
-			},
-		})
-		if err != nil {
-			a.logger.Error("failed to get task list backlog", tag.Error(err))
-			break
+	result := make(map[int]*types.TaskListPartition, targetWritePartitions)
+
+	for i := 0; i < targetWritePartitions; i++ {
+		if p, ok := writePartitions[i]; ok {
+			result[i] = p
+		} else {
+			result[i] = &types.TaskListPartition{}
 		}
-		nw := int32(1)
-		if resp.PartitionConfig != nil {
-			nw = resp.PartitionConfig.NumWritePartitions
+	}
+	return result, true
+}
+
+func (a *adaptiveScalerImpl) adjustReadPartitions(m *aggregatePartitionMetrics, oldReadPartitions map[int]*types.TaskListPartition, newWritePartitions map[int]*types.TaskListPartition) (map[int]*types.TaskListPartition, bool) {
+	result := make(map[int]*types.TaskListPartition, len(newWritePartitions))
+	changed := false
+	for id, p := range oldReadPartitions {
+		result[id] = p
+	}
+	for id, p := range newWritePartitions {
+		if _, ok := result[id]; !ok {
+			changed = true
 		}
-		// in order to drain a partition, 2 conditions need to be met:
-		// 1. the backlog size is 0
-		// 2. no task is being added to the partition, which is guaranteed to be true if the partition knows that the number of write partition is less or equal to its partition ID
-		if resp.TaskListStatus.GetBacklogCountHint() == 0 && nw <= i {
-			// if the partition is drained, we can downscale the number of read partitions
-			numReadPartitions = i
+		result[id] = p
+	}
+
+	for i := len(result) - 1; i >= len(newWritePartitions); i-- {
+		if m.byPartition[i].isDrained() {
 			changed = true
+			delete(result, i)
 		} else {
 			break
 		}
 	}
 	if changed {
-		a.logger.Info("adjust read partitions", tag.NumReadPartitions(numReadPartitions), tag.NumWritePartitions(numWritePartitions))
+		a.logger.Info("adjust read partitions", tag.NumReadPartitions(len(result)), tag.NumWritePartitions(len(newWritePartitions)))
+	}
+	return result, changed
+}
+
+func (a *adaptiveScalerImpl) collectPartitionMetrics(config *types.TaskListPartitionConfig) (*aggregatePartitionMetrics, error) {
```

Comment:
This might increase the load significantly. Originally, describeTaskList is only called when the number of write partitions is less than the number of read partitions. But after this change, it will be called periodically. Maybe, we can add a metric to track how much extra load is caused by this.

Reply:
Good point. Failure case behavior is also not clear to me. If one partition fails to response (e.g. timeout) what do we do? 
Maybe we just activate this new logic if the tasklist has isolation groups enabled. This way we can iterate and optimize as needed without impacting normal behavior of adaptive scaler

Reply:
Failure case is a no-op, which seems reasonable as a starting point. In the future we can explore more sophisticated approaches such as maintaining more state. I've updated this logic to only perform the RPC to all child partitions when isolation is enabled.